AI/MMAI/BAI/v13/stack.cpp

```diff
@@ -0,0 +1,543 @@
+		}
+		else
+		{
+			// "special" slot, e.g. summoned, commander, etc.
+			return S13::STACK_SLOT_SPECIAL;
+		}
+	}
+
+	char calculateAlias(int slot)
+	{
+		switch(slot)
+		{
+			case S13::STACK_SLOT_SPECIAL:
+				return 'S';
+				break;
+			case S13::STACK_SLOT_WARMACHINES:
+				return 'M';
+				break;
+			default:
+				return static_cast<char>('0' + slot);
+		}
+	}
+}
+
+// static
+int Stack::CalcValue(const CCreature * cr)
+{
+	static const bool verbose = []()
+	{
+		const char * envvar = std::getenv("MMAI_VERBOSE");
+		return envvar != nullptr && std::strcmp(envvar, "1") == 0;
+	}();
+
+	static std::map<const CCreature *, float> cache;
+	static std::mutex cache_mutex;
+
+	{
+		std::lock_guard lock(cache_mutex);
+		auto it = cache.find(cr);
+		if(it != cache.end())
+			return static_cast<int>(it->second);
+	}
+
+	/*
+	 * Formula:
+	 * 10 * (A + B) * C * D1 * D2 * ... * Dn
+	 *
+	 * A = <offensive factor>
+	 * B = <defensive factor>
+	 * C = <speed factor>
+	 * D* = <bonus factor>
+	 */
+
+	auto att = cr->getBaseAttack();
+	auto def = cr->getBaseDefense();
+	auto dmg = (cr->getBaseDamageMax() + cr->getBaseDamageMin()) / 2.0;
+	auto hp = cr->getBaseHitPoints();
+	auto spd = cr->getBaseSpeed();
+	auto shooter = cr->hasBonusOfType(BonusType::SHOOTER);
+	auto bonuses = cr->getAllBonuses(Selector::all);
+
+	auto a = 3 * dmg * (1 + std::min(4.0, 0.05 * att));
+	auto b = hp / (1 - std::min(0.7, 0.025 * def));
+	auto c = spd ? std::log(spd * 2) : 0.5;
+	auto d = shooter ? 1.5 : 1.0;
+
+	for(const auto & bonus : *bonuses)
+	{
+		switch(bonus->type)
+		{
+			case BonusType::ADDITIONAL_ATTACK:
+				d += (shooter ? 0.5 : 0.3);
+				break;
+			case BonusType::ADDITIONAL_RETALIATION:
+				d += (bonus->val * 0.1);
+				break;
+			case BonusType::ATTACKS_ALL_ADJACENT:
+				d += 0.2;
+				break;
+			case BonusType::BLOCKS_RETALIATION:
+				d += 0.3;
+				break;
+			case BonusType::DEATH_STARE:
+				d += (bonus->val * 0.02); // 10% = 0.2
+				break;
+			case BonusType::DOUBLE_DAMAGE_CHANCE:
+				d += (bonus->val * 0.005); // 20% = 0.1
+				break;
+			case BonusType::ENEMY_DEFENCE_REDUCTION:
+				d += (bonus->val * 0.0025); // 40% = 0.1
+				break;
+			case BonusType::FIRE_SHIELD:
+				d += (bonus->val * 0.003); // 20% = 0.1
+				break;
+			case BonusType::FLYING:
+				d += 0.1;
+				break;
+			case BonusType::LIFE_DRAIN:
+				d += (bonus->val * 0.003); // 100% = 0.3
+				break;
+			case BonusType::NO_DISTANCE_PENALTY:
+				d += 0.5;
+				break;
+			case BonusType::NO_MELEE_PENALTY:
+				d += 0.1;
+				break;
+			case BonusType::THREE_HEADED_ATTACK:
+				d += 0.05;
+				break;
+			case BonusType::TWO_HEX_ATTACK_BREATH:
+				d += 0.1;
+				break;
+			case BonusType::UNLIMITED_RETALIATIONS:
+				d += 0.2;
+				break;
+			case BonusType::SPELL_LIKE_ATTACK:
+				switch(bonus->subtype.as<SpellID>())
+				{
+					case SpellID::DEATH_CLOUD:
+						d += 0.2;
+				}
+				break;
+			case BonusType::SPELL_AFTER_ATTACK:
+				switch(bonus->subtype.as<SpellID>())
+				{
```

Comment:
Would have been great to somehow cover units from mods, at least with some default value